mh/session.py

```diff
@@ -139,6 +139,24 @@ def get_server(self):
         assert self.local_info['server_id'] is not None
         return DB.get_server(self.local_info['server_id'])
 
+    def get_server_full(self, server_id):
+        return DB.get_server(server_id).is_full()
+
+    def get_gate_full(self, server_id, gate_id):
+        return DB.get_gate(server_id, gate_id).is_full()
+
+    def get_city_full(self, server_id, gate_id, city_id):
+        return DB.get_city(server_id, gate_id, city_id).is_full()
+
+    def get_city_empty(self, server_id, gate_id, city_id):
+        return DB.get_city(server_id, gate_id, city_id).is_empty()
+
+    def find_gate(self, server_id, gate_id):
+        return DB.get_gate(server_id, gate_id)
+
+    def find_city(self, server_id, gate_id, city_id):
+        return DB.get_city(server_id, gate_id, city_id)
```

Comment:
I'm really not a huge fan of this since the `session` module is mainly used to cache the server data and hide the implementation logic _(for instance which DB backend we use, which function to call based on the packet we received and its parameters, etc.)_. That's also why most of its functions don't ask for the `server_id` as parameter and use the `local_info` when needed rather than being a simple wrapper to call the default database instance.

I'll let it slide for now since we don't provide a better abstraction to query the database directly, especially for other servers. AFAICT, this point seems still valid with the `state` abstraction you proposed in another PR because parts of the `session` modules (including the ones this PR added) needs to take into account whether or not the city/gate/server is handled by the current server process or not.